### DISCUSSION TITLE
issue: Items are not correctly removed from field array

### DISCUSSION THREAD

**Discussion by alexanderkogan** [react-hook-form/react-hook-form#9496] [Q&A] [answered]
### Version Number

7.31.1

### Codesandbox/Expo snack

https://codesandbox.io/s/fieldarraytablevirtuoso-chl9cs

### Steps to reproduce

1. Open the Console
2. Click on 'Append'
3. Click on 'Send'
4. See an array of length 2 printed out.
5. Click on 'Delete'
6. Click on 'Send'
7. See an array of length 2 erroneously printed out.

### Expected behaviour

The onSubmit function should receive an array of length 1, the array seems to be stale though.
Without TableVirtuoso the array behaves as expected, see: https://codesandbox.io/s/fieldarraymap-8f3x6y

### What browsers are you seeing the problem on?

Firefox, Chrome

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by bluebill1049** [comment]
Not an issue with hook form, this TableVirtuoso library is not compatible with hook form.

**Comment by leapful** [comment, marked as answer]
@alexanderkogan @bluebill1049 

The main issue is from library react-virtuoso which is using hooks to extract and memorize prop `data` (list of item) for root TableVirtuoso component.

Whenever prop `data` is changed, for example removing one item, React will trigger re-rendering on TableVirtuoso, but at that time, hook to extract new data is not executed yet therefore old data is still being used and perform registering old field array values again with Controller component (your `text-field.tsx` file) of react-hook-form. 

A workaround is to check if field array item is already removed to avoid re-rendering Controller component.

Working demo: https://codesandbox.io/s/fieldarraytablevirtuoso-forked-rgxfv1?file=/src/App.tsx

**Reply by bluebill1049** [reply]
thanks @leapful for looking into this for us.

**Reply by alexanderkogan** [reply]
Thank you for the workaround. Now only the last item is removed though, because `isRemoved` is only checking on the index. It has to check on some unique identifier, I guess.
